### ISSUE TITLE
Integer overflows in unbounded integer Box spaces

### ISSUE THREAD

**Issue by iamlucaswolf** [openai/gym#1929]
Hi,

I believe I noticed some peculiar bugs in `gym.spaces.Box` around the representation of integer box spaces with unbounded upper ends (e.g. [0, inf)).

First of all, instantiating a Box space with an upper bound of `np.inf` will immediately result in an integer overflow:

```python
>>> space = Box(low=0, high=np.inf, shape=(1,), dtype=np.int)
>>> space.high
array([-9223372036854775808])
```

This is due to line 45f in `gym/spaces/box.py`, which broadcasts the upper bound to all dimensions in the constructed space if the input argument `high` is scalar. The implementation uses `np.full`, which will cause the overflow. 

```python
>>> np.full((1,), np.inf, dtype=np.int)
array([-9223372036854775808])
```

While it's certainly debatable whether these are the intended semantics of float to int conversion (after all, `type(np.inf) == float`), there exists an easy workaround for this issue...

```python
>>> space = Box(low=0, high=np.iinfo(np.int).max, shape=(1,), dtype=np.int)
>>> space.high
array([9223372036854775807])
```

... which will however result in more problems down the line. If we now try to sample from this space, we get:

```python
>>> space = Box(low=0, high=np.iinfo(np.int).max, shape=(1,), dtype=np.int)
>>> space.sample()
array([-8746956029393751040])
```

In this case, the overflow happens in line 95 (in `Box.sample()`), which sets `high = self.high.astype('int64') + 1`; `high` is then used as an (exclusive) upper bound for uniform sampling in 117f.

**TL;DR**

Currently, it seems that integer Box spaces with unbounded upper endpoints can not be represented in `gym`. Trying to do so will result in integer overflows which happen silently along the way.

While I agree that this is a somewhat pathological case, I'd argue that it does have practical implications. I stumbled over it while implementing an `env` class for chess with the board encoding used by AlphaZero.

I'd be happy to provide a PR for this.

**Comment by christopherhesse** [comment]
Thanks for the bug report!   I don't see the use case for unbounded upper endpoints for integers, which have bounded upper end points.  I agree that the high + 1 stuff is an issue, but I think it might work with a special case for int64/uint64, since all smaller integers would fit in those.

What sort of PR are you envisioning there?

**Comment by iamlucaswolf** [comment]
Thanks for your reply!

**Integers being bounded**: I see your point, but aren't these two pair of shoes? Integers are implicitly bounded due to the number of bits representing them – but so are floats; IEEE754 specifying reserved values to represent +/- infinity is just a design decision. From a standpoint of pure semantics, I see no reason why $\mathbb{N}$ should be less unbounded than $\mathbb{R}$.

**Use cases**: I would argue it makes sense to have the ability to model integer counters (such as timers, game scores, etc). These need not assume crazy high values necessarily, but sometimes an upper bound might just not be apparent or easy to determine from the application context alone. As mentioned above, I recently encountered such a case when writing [a gym env for chess](https://github.com/iamlucaswolf/gym-chess/blob/master/gym_chess/alphazero/board_encoding.py) that implements the AlphaZero board encoding.

**PR**: Depends on what semantics you consider best. If integers are bounded, adding some special cases should do the trick, as you already mentioned. However, if integers are unbounded, sampling from a uniform U(k, max int) makes little sense, as U(k, infinity) is undefined. Perhaps in that case sampling from a shifted geometric would be more appropriate, since we sample from a shifted exponential in the continuous case?

**Comment by christopherhesse** [comment]
So if I understand what you're saying correctly, you want:

* sampling from integer box spaces where high=maxint(int64) doesn't work.  I think this is related to https://github.com/openai/gym/issues/1256, where this previously didn't work for `uint8`.  If I had to guess the issue could be related to `sample()` using `random.uniform()` instead of `random.randint()`.  In order to specify a high value of 2**64, it looks like you actually have to use a `dtype` of `object` to use python long ints, since a uint64 will max out at 2**64-1.

* a high value of `np.inf` should be supported for integer spaces.  I would prefer if high/low arrays have the same dtype as the space to avoid confusion, so I don't think `np.inf` makes sense there, unless it's immediately converted to the max of the int type, and even then I think users could use `np.iinfo(np.int).max` as you have shown.

Is that correct?

**Comment by RedTachyon** [comment]
So it's been a while since this was brought up, and while the priority on this isn't necessarily very high, I think it'd make sense to change the bounds on int boxes. 

It might be tricky to have a proper integer infinity. It might be an arbitrary design from IEEE754, but that's what we're working with. The least bad option imo would be just casting `np.inf` to the maximum/minimum values of the relevant datatype. (maybe one less than maximum and one more than minimum, to avoid the sampling problem)

But again, this is imo relatively low priority, because using `np.inf` to set bounds on an integer space is basically asking for trouble - even if it makes some intuitive sense.
